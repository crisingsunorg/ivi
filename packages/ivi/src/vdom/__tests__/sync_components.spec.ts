describe(`sync components`, () => {
  /* tslint:disable:whitespace */
  let ivi: typeof import("ivi");
  let html: typeof import("ivi-html");
  let utils: typeof import("./utils");
  /* tslint:enable:whitespace */

  beforeEach(async () => {
    jest.resetModules();
    ivi = await import("ivi");
    html = await import("ivi-html");
    utils = await import("./utils");
  });

  test(`#1`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          html.span()
        );
        const v2 = (
          utils.Stateful(
            html.div(),
          )
        );

        r(v1);
        const n = r(v2);

        expect(n).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#2`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          utils.Stateful(
            html.div(),
          )
        );
        const v2 = (
          html.div()
        );

        const a = r(v1);
        const b = r(v2);

        expect(a).not.toBe(b);
        expect(b).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#3`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          html.div()
        );
        const v2 = (
          utils.Stateful(
            html.div(),
          )
        );

        const a = r(v1);
        const b = r(v2);

        expect(a).not.toBe(b);
        expect(b).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#4`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          utils.Stateful(
            html.div(),
          )
        );
        const v2 = (
          html.span()
        );

        r(v1);
        const n = r(v2);

        expect(n).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#5`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          utils.Stateful(
            html.div(),
          )
        );
        const v2 = (
          utils.Stateful(
            html.div(),
          )
        );

        const a = r(v1);
        const b = r(v2);

        expect(a).toBe(b);
        expect(b).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#6`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          utils.Stateful(
            utils.Stateful(
              html.div(),
            ),
          )
        );
        const v2 = (
          html.span()
        );

        r(v1);
        const b = r(v2);

        expect(b).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#7`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          html.span()
        );
        const v2 = (
          utils.Stateful(
            utils.Stateful(
              html.div(),
            ),
          )
        );

        r(v1);
        const b = r(v2);

        expect(b).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`#8`, () => {
    utils.startRender(r => {
      utils.checkDOMOps(c => {
        const v1 = (
          utils.Stateful(
            utils.Stateful(
              html.div(),
            ),
          )
        );
        const v2 = (
          utils.Stateful(
            utils.Stateful(
              html.div(),
            ),
          )
        );

        const a = r(v1);
        const b = r(v2);

        expect(a).toBe(b);
        expect(b).toMatchSnapshot();
        expect(c).toMatchSnapshot();
      });
    });
  });

  test(`component should raise an exception when render function returns children collection`, () => {
    utils.startRender<HTMLElement>(r => {
      const v1 = (
        utils.Stateful(
          html.div(),
        )
      );
      const v2 = (
        utils.Stateful(
          ivi.fragment(
            html.div(),
            html.div(),
          )!,
        )
      );

      r(v1);
      expect(() => { r(v2); }).toThrowError("singular");
    });
  });
});
