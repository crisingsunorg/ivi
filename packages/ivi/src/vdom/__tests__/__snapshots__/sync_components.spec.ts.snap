// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sync components #1 1`] = `<div />`;

exports[`sync components #1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 1,
}
`;

exports[`sync components #2 1`] = `<div />`;

exports[`sync components #2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 1,
}
`;

exports[`sync components #3 1`] = `<div />`;

exports[`sync components #3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 1,
}
`;

exports[`sync components #4 1`] = `<span />`;

exports[`sync components #4 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 1,
}
`;

exports[`sync components #5 1`] = `<div />`;

exports[`sync components #5 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 0,
}
`;

exports[`sync components #6 1`] = `<span />`;

exports[`sync components #6 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 1,
}
`;

exports[`sync components #7 1`] = `<div />`;

exports[`sync components #7 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 1,
}
`;

exports[`sync components #8 1`] = `<div />`;

exports[`sync components #8 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "insertBefore": 1,
  "removeChild": 0,
  "replaceChild": 0,
}
`;
